tool/tctl/common/auth_command.go

```diff
@@ -539,6 +539,8 @@ func hostCAFormat(ca services.CertAuthority, keyBytes []byte, client auth.Client
 	if err != nil {
 		return "", trace.Wrap(err)
 	}
-	allowedLogins, _ := roles.CheckLoginDuration(defaults.MinCertDuration + time.Second)
+	// RiskyPermitEmptyLogins is OK here because we aren't using CheckLoginDuration to calculate
+	// certificate generation parameters.
+	allowedLogins, _ := roles.CheckLoginDuration(defaults.MinCertDuration+time.Second, services.RiskyPermitEmptyLogins(true))
```

Comment:
Based off the comment, I don't think it warrants changes in the `CheckLoginDuration` as the function's purpose is slightly different and I feel this is mixing a couple of concepts together - compute a set of logins based on ttl and validate the logins.
I'd suggest factoring out the query of logins based on ttl out of `CheckLoginDuration`:
```go
// CheckLoginDuration checks if role set can login up to given duration and
// returns a combined list of allowed logins.
func (set RoleSet) CheckLoginDuration(ttl time.Duration) ([]string, error) {
	logins, matched := set.LoginsWithTTL(ttl)
	if !matched {
		return nil, trace.AccessDenied("this user cannot request a certificate for %v", ttl)
	}
	if len(logins) == 0 && !set.hasPossibleLogins() {
		// user was deliberately configured to have no login capability,
		// but ssh certificates must contain at least one valid principal.
		// we add a single distinctive value which should be unique, and
		// will never be a valid unix login (due to leading '-').
		logins = append(logins, "-teleport-nologin-"+uuid.New())
	}
	if len(logins) == 0 {
		return nil, trace.AccessDenied("this user cannot create SSH sessions, has no allowed logins")
	}
	return logins, nil
}

// LoginsWithTTL returns the list of logins that match the specified ttl value.
// If ttl is 0, all logins are returned
// matched indicates whether ttl matched a role even though it might not have contained allowed logins
func (set RoleSet) LoginsWithTTL(ttl time.Duration) (logins []string, matched bool) {
	for _, role := range set {
		maxSessionTTL := role.GetOptions().MaxSessionTTL.Value()
		if ttl <= maxSessionTTL && maxSessionTTL != 0 {
			matched = true
			for _, login := range role.GetLogins(Allow) {
				logins = append(logins, login)
			}
		}
	}
	return logins, matched
}
```
and then it could be used directly here:
```go
// ...
allowedLogins, _ := roles.LoginsWithTTL(defaults.MinCertDuration + time.Second)
return sshutils.MarshalAuthorizedHostsFormat(ca.GetClusterName(), keyBytes, allowedLogins)
```
`LoginsWithTTL` does not have to be part of the `AccessChecker` interface.

Reply:
Can't do something much simpler here and just call `roles.CheckLoginDuration` then filter out any logins that start with `-`?